### PR TITLE
[AV][Android] Fix AV progress update

### LINKED ISSUE #6342: [android][video] The more you replay a video, the more onPlaybackStatusUpdate will be called
### Environment

```
  Expo CLI 3.4.1 environment info:
    System:
      OS: Linux 4.15 Ubuntu 18.04.3 LTS (Bionic Beaver)
      Shell: 4.4.20 - /bin/bash
    Binaries:
      Node: 12.10.0 - ~/.nvm/versions/node/v12.10.0/bin/node
      Yarn: 1.17.3 - ~/.nvm/versions/node/v12.10.0/bin/yarn
      npm: 6.10.3 - ~/.nvm/versions/node/v12.10.0/bin/npm
    IDEs:
      Android Studio: 3.5 AI-191.8026.42.35.5791312
    npmPackages:
      @types/react: ^16.8.23 => 16.9.11 
      @types/react-native: ^0.57.65 => 0.57.65 
      expo: ^35.0.0 => 35.0.0 
      react: 16.8.3 => 16.8.3 
      react-native: https://github.com/expo/react-native/archive/sdk-35.0.0.tar.gz => 0.59.8 
      react-navigation: ^4.0.10 => 4.0.10 
    npmGlobalPackages:
      expo-cli: 3.4.1

```

Target app is android, I test on Emulator Android 9.

### Steps to Reproduce

Create a new app, take a video from the internet and just use my code.  
The console will be flooded by numbers which will increase quicker and quicker, the more you click on the replay button.

### Expected Behavior

The `onPlaybackStatusUpdate` calls should be consistent over video replay

### Actual Behavior

Each time you click on replay button which call playAsync or replayAsync, the callback is called one more time per update.  
If you push the button 10 times, instead of calling the `onPlaybackStatusUpdate` 1 time every 500ms, it will be called 10 times every 500ms.

### Reproducible Demo

```
class VideoV extends React.Component {
  constructor (props) {
    super(props)
    this._replayVideo = this._replayVideo.bind(this)
    this.count = 0
  }

  _handleVideoRef = component => {
    this.playbackObject = component
  }

  _replayVideo () {
    this.playbackObject.getStatusAsync().then(({ positionMillis, durationMillis }) => {
      console.log(Object.keys(this.playbackObject))
      if (positionMillis === durationMillis) {
        this.playbackObject.playAsync()
      }
    })
  }

  render () {
    return (
      <View style={{ height: 500, width: 500 }}>
        <Video
          style={{ width: '80%', height: '80%' }}
          source={require('yourLocalVideoUrl.mp4')}
          shouldPlay
          ref={this._handleVideoRef}
          onPlaybackStatusUpdate={(params) => {
            this.count++
            console.log(this.count)
          }}
        />
        <TouchableOpacity onPress={this._replayVideo} style={{ height: 200, width: 200 }}>
          <Text style={{ color: 'white' }}>
            replay
          </Text>
        </TouchableOpacity>
      </View>
    )
  }
}
```

Please let me know if I can provide more details

### PR DESCRIPTION
**Pull request by mczernek**
# Why

Resolves #6342

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,18 @@ This is the log of notable changes to the Expo client that are developer-facing.
...
 
+### 🛠 Breaking changes
+
+### 🎉 New features
+
+### 🐛 Bug fixes
+
+- Fixed multiplied callbacks in av after replaying ([#7193](https://github.com/expo/expo/pull/7193) by [@mczernek](https://github.com/mczernek))
+
+## 37.0.0
```
Comment (bbarthec) [inline CHANGELOG.md]: IIRC adding new SDK section is done in release cycle.
Please remove it from here 😊

File: android/build.gradle [hunk 2]
```diff
@@ -9,7 +9,7 @@ buildscript {
     dbFlowVersion = '4.2.4'
     buildToolsVersion = '28.0.0'
     supportLibVersion = '28.0.0'
-    kotlinVersion = '1.3.50'
+    kotlinVersion = '1.3.61'
```
Comment (bbarthec) [inline android/build.gradle]: Should be done in separate PR IMO 🤔 And I wish it would be alongside some other libs/deps update 😉

File: packages/expo-av/android/build.gradle [hunk 3]
```diff
@@ -61,6 +72,13 @@ if (new File(rootProject.projectDir.parentFile, 'package.json').exists()) {
 }
 
 dependencies {
+
+//  Uncomment to run unit tests
+//  implementation 'org.junit.jupiter:junit-jupiter-api:5.5.1'
+//  runtimeOnly "org.junit.jupiter:junit-jupiter-engine:5.5.1"
+//
+//  implementation 'io.mockk:mockk:1.9'
```
Comment (bbarthec) [inline packages/expo-av/android/build.gradle]: Why have you done it like this? 🤔 Cannot tests libs be always included?

File: packages/expo-av/android/src/main/java/expo/modules/av/progress/AndroidLooperTimeMachine.kt [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+class AndroidLooperTimeMachine: TimeMachine {
+
+  override fun scheduleAt(intervalMillis: Long, callback: TimeMachineTick) {
+    Handler().postDelayed(callback, intervalMillis)
+  }
+
+  override val time: Long
+    get() = System.currentTimeMillis()
+}
```
Comment (bbarthec) [inline packages/expo-av/android/src/main/java/expo/modules/av/progress/AndroidLooperTimeMachine.kt]: newline please 🙏

File: packages/expo-av/android/src/main/java/expo/modules/av/progress/ProgressLooper.kt [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+  private fun calculateNextInterval() =
+    if (nextExpectedTick > timeMachine.time) {
+      interval
+    } else {
+      (((timeMachine.time - nextExpectedTick) / interval) + 1) * interval
+    }
+
+
+}
```
Comment (bbarthec) [inline packages/expo-av/android/src/main/java/expo/modules/av/progress/ProgressLooper.kt]: please configure your AS to add newlines automagically 🙇

File: packages/expo-av/android/src/test/java/expo/modules/av/progress/ProgressLooperTest.kt [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+//    timeMachine.triggerListeners()
+//    verify(exactly = 2) { timeMachine.scheduleAt(any(), any()) }
+//
+//    looper.stopLooping()
+//    timeMachine.advanceBy(100000)
+//    timeMachine.triggerListeners()
+//    verify(exactly = 2) { timeMachine.scheduleAt(any(), any()) }
+//  }
+//
+//}
```
Comment (bbarthec) [inline packages/expo-av/android/src/test/java/expo/modules/av/progress/ProgressLooperTest.kt]: newline

Comment (bbarthec) [inline packages/expo-av/android/src/test/java/expo/modules/av/progress/ProgressLooperTest.kt]: I don't get the idea of commenting out whole test file? 🤔 
Cannot it be done some other (more programmer-friendly) way?

File: packages/expo-av/android/src/main/java/expo/modules/av/progress/ProgressLooper.kt [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+  fun scheduleAt(intervalMillis: Long, callback: TimeMachineTick)
+  val time: Long
+}
+
+/**
+ * @mczernek: NOTE that this class has corresponding test suite. However, since I was unable to set
+ * them up correctly, in order to run them, you should uncomment dependencies in module's build.gradle,
+ * and contents of test class: at packages/expo-av/android/src/test/java/expo/modules/av/progress/ProgressLooperTest.kt
+ * Note, that this breaks client's build, so remember to comment it back.q
+ */
```
Comment (bbarthec) [inline packages/expo-av/android/src/main/java/expo/modules/av/progress/ProgressLooper.kt]: Is it possible to extract tests to other PR where we can discuss what's wrong with the tests? 🤔

File: packages/expo-av/android/src/main/java/expo/modules/av/progress/ProgressLooper.kt [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+      if (!value) {
+        interval = 0L
+        nextExpectedTick = -1L
+        waiting = false
+      }
+    }
+
+  private var listener: PlayerProgressListener? = null
+
+  fun changeListener(listener: PlayerProgressListener) {
```
Comment (bbarthec) [inline packages/expo-av/android/src/main/java/expo/modules/av/progress/ProgressLooper.kt]:
```suggestion
  fun setListener(listener: PlayerProgressListener) {
```
I believe this is more Android-style 🤔
[Here View example](https://developer.android.com/reference/android/view/View?hl=pt-br#setOnApplyWindowInsetsListener(android.view.View.OnApplyWindowInsetsListener )

File: packages/expo-av/android/src/main/java/expo/modules/av/progress/ProgressLooper.kt [hunk 9]
```diff
@@ -0,0 +1,74 @@
+package expo.modules.av.progress
+
+typealias TimeMachineTick = () -> Unit
+
+interface TimeMachine {
+  fun scheduleAt(intervalMillis: Long, callback: TimeMachineTick)
```
Comment (bbarthec) [inline packages/expo-av/android/src/main/java/expo/modules/av/progress/ProgressLooper.kt]: I was wondering why have you chosen such name and parameters order, but it's to allow such syntax:?
```kt
timeMachine.scheduleAt(nextExpectedTick - timeMachine.time) {
  waiting = false
  listener?.invoke()
  scheduleNextTick()
}
```

File: packages/expo-av/android/src/main/java/expo/modules/av/progress/ProgressLooper.kt [hunk 10]
```diff
@@ -0,0 +1,74 @@
+package expo.modules.av.progress
+
+typealias TimeMachineTick = () -> Unit
+
+interface TimeMachine {
```
Comment (bbarthec) [inline packages/expo-av/android/src/main/java/expo/modules/av/progress/ProgressLooper.kt]: I don't like this name TBH 😛
What `TimeMachine` means?
I dunno what might be better, but maybe sth that connects with JS's `Timer`/`Timeout` or Java's/Kotlin's `Handler` or `Timer` from [this SO answer](https://stackoverflow.com/a/43348715/4337317)

**Review by bbarthec — CHANGES_REQUESTED** [review]
I wish we don't embed maintenance changes into PRs that solves issues 😉

File: packages/expo-av/android/src/main/java/expo/modules/av/progress/ProgressLooper.kt (hunk 7 above)
Comment (mczernek) [reply packages/expo-av/android/src/main/java/expo/modules/av/progress/ProgressLooper.kt]: I've already solved this issue. It's coming.

File: packages/expo-av/android/src/main/java/expo/modules/av/progress/ProgressLooper.kt (hunk 9 above)
Comment (mczernek) [reply packages/expo-av/android/src/main/java/expo/modules/av/progress/ProgressLooper.kt]: I generally like to put callbacks at the end. Such syntax is one of main reasons. Another is, that I find this code hard to read:
```
new Handler.postDelayed(() -> {
  // some code here
  // possibly multiple lines
}, 500); // And here the actual delay.
```

File: packages/expo-av/android/src/main/java/expo/modules/av/progress/ProgressLooper.kt (hunk 10 above)
Comment (mczernek) [reply packages/expo-av/android/src/main/java/expo/modules/av/progress/ProgressLooper.kt]: It advences you in time by a given period. What better name for that?

File: packages/expo-av/android/src/main/java/expo/modules/av/progress/ProgressLooper.kt (hunk 8 above)
Comment (mczernek) [reply packages/expo-av/android/src/main/java/expo/modules/av/progress/ProgressLooper.kt]: TBH, I don't know. What I wanted to emphasize here is the fact, that we actually change the listener, not set another one.
I know there is distinction between `set` and `add` which makes just that, so I'll go with that if current name confuses you.

File: CHANGELOG.md [hunk 11]
```diff
@@ -16,6 +16,7 @@ This is the log of notable changes to the Expo client that are developer-facing.
 
 ### 🐛 Bug fixes
 
+- Fixed multiplied callbacks in av after replaying ([#7193](https://github.com/expo/expo/pull/7193) by [@mczernek](https://github.com/mczernek))
```
Comment (bbarthec) [inline CHANGELOG.md]:
```suggestion
- Fixed multiplied callbacks in `expo-av` after replaying ([#7193](https://github.com/expo/expo/pull/7193) by [@mczernek](https://github.com/mczernek))
```

**Review by bbarthec — APPROVED** [review]
LGTM 💯 only this one small change in `CHANGELOG` 💅 😉

**Review by bbarthec — APPROVED** [review]
